### ISSUE TITLE
Move from Python2 to Python3

### ISSUE THREAD

**Issue by Foorack** [EFForg/https-everywhere#6749]

**Comment by capt2101akash** [comment]
Can you please describe the whole issue regarding this bug

**Comment by Foorack** [comment]
@capt2101akash Will explain more in detail later today. Essentially you are unable to push if you uninstall python2.7. During push git runs the tests which currently require python 2.7.

**Comment by Foorack** [comment]
@capt2101akash Hello again. Sorry for the delay.

First off I would like to clarify this is not a issue but rather an improvement request.
For more information what I am referring to:
https://github.com/EFForg/https-everywhere/blob/master/test/rules.sh#L30
As one can see it is currently hard-coded to fail if python2.7 is not installed on the system.

**Comment by jeremyn** [comment]
It's possible that the `python2.7` requirement in `rules.sh` is unnecessary, for example `check_rules.py` in `fetch.sh` [here](https://github.com/EFForg/https-everywhere/blob/master/test/fetch.sh#L22) just executes with `python`. On the other hand, `fetch.sh` is used mainly for Travis testing and it's possible for simplicity in the script that we just assume the environment uses 2.7, since we control that environment.

@fuglede @Hainish @J0WI Can any of you give us any information you have about Python 2 vs Python 3 support for this project?

**Comment by Hainish** [comment]
This definitely needs to happen at some point, since Python 2.7 is being [EOLed in 2020](https://pythonclock.org/).  However this is not a priority for us currently, as other development requirements (such as the transition to WebExtensions for Firefox) take precedent.

**Comment by jeremyn** [comment]
If `python2.7` is actually required then I agree that it's a bug anywhere that we refer to `python` or `/usr/bin/python`.

**Comment by renyouguo** [comment]
Python3 support unicode by default, may fix the encoder errors just like: https://travis-ci.org/EFForg/https-everywhere/jobs/161551245#L358

**Comment by cowlicks** [comment]
Partial fixes in #12996

**Comment by Hainish** [comment]
# Python 2 to 3 Migration Work

I've been working on modifying [`zipfile_determinstic.py`](https://github.com/Hainish/https-everywhere/blob/ec42ff8e0ceefbc78807b0d40c9e45e528c5fbaa/utils/zipfile_deterministic.py) (which is sourced from the python standard library) https://github.com/EFForg/https-everywhere/pull/14515 and have it successfully building the extension in python 3.6 deterministically.  I also think we should move forward with #12996 and other efforts which are more specific about the python version supported.

## Python 3.6

### The Bleeding-Edge Problem 

#14515 also adds python 3.6 as a hard dependency.  The intention is to slowly migrate our python infrastructure from 2.7 to 3.6, while still supporting both during a transition period.  The reasoning behind this is that python 3.6 is the latest stable python, and as such will be supported for the longest amount of time into the future (3.7 is expected to hit stable [in June](https://www.python.org/dev/peps/pep-0537/)).  However, amongst the maintained stable releases for Ubuntu and Debian, only Ubuntu 17.10 provides python 3.6 by default from the package manager.  It will not be available to most developers just yet without building it or relying on a separate OS image.  This means that we would be adding a dependency that hardly anyone has yet.  However, it is likely that by choosing a stable but new release of python 3, we'll give ourselves enough runway time so that we can fully  support it when it is deployed to the major development distributions.

### Possible Resolutions

In #13815, there was discussion on what OSes we want to support.  If we move to a model where we specify a docker image as a development dependency, it will be easier to justify a transition to 3.6: we can swap the development image relatively easily.  However, I think this warrants a certain amount of discussion and reasonable planning.  I don't want to disrupt developers workflow here unnecessarily, but I'd also like to maintain the flexibility and prerogative to make a versioning decision once fruitful feedback has been exhausted.

Since this is a more specific plan but still involves discussion on python 2 to 3 migration, I've decided to open the discussion here rather than in a new issue.  I've also added the "question" label to indicate that feedback is requested.

**Comment by Hainish** [comment]
In the meantime I will continue working on python 3.6 support in a separate branch and rebasing from master where appropriate.  Rebasing is additional work I'd like to avoid moving forward, so I'd like to prioritize merging python 3 support into master, contingent on this conversation.

**Comment by jeremyn** [comment]
I assume the problem `zipfile_deterministic.py` is intended to solve is to allow us to create the zip file for the extension build so that it has the same hash every time, since regular zipping can introduce differences.

It seems like it would be very convenient if this technical problem could be offloaded into some other library or program. It is not a core competence of HTTPS Everywhere. It obviously requires ongoing maintenance. It also, since it derives from PSF code, introduces some of the licensing problems discussed in issue https://github.com/EFForg/https-everywhere/issues/7978.

@Hainish Have you made a recent search for some replacement for this library/program? It has been years since it was introduced into the build process. The Python ecosystem has grown, and we could use Node stuff too.

**Comment by Hainish** [comment]
@jeremyn Python has not introduced determinism into the `zipfile` module.  There does seem to be a node module [out there](https://github.com/bitgenics/deterministic-zip), but I worry about relying on yet another dependency which is likely to be abandoned.  Sourcing in the `zipfile` standard library and removing the sources of entropy seems like a reasonable long-term solution, since this will only have to be modified once every time we switch supported versions of python.  By that time, there may be more reliable ways to assure determinism.

The PSF code is GPL-compatible.  The fact that we have a LICENSE.txt document which supports the flexibility of multiple licenses is fortunate, and a discussion that has already been made.

I'd like to focus this conversation on the overall pathway to python 3, rather than specific scripts and implementation details.

**Comment by jeremyn** [comment]
@Hainish I agree that https://github.com/bitgenics/deterministic-zip appears unsuitable. There may be other ways that will work for us, see for example https://wiki.debian.org/ReproducibleBuilds/TimestampsInZip . However since you specifically don't want to discuss this, I'll stop there.

Are you asking how many contributors would be seriously inconvenienced by requiring 3.6 for builds? In that case you might just ask that question, in here or https://github.com/EFForg/https-everywhere/issues/13815 ("Which developer operating systems do we support?"), and ping the maintainers and top 10 or so contributors for a thumbs-up/thumbs-down.

**Comment by J0WI** [comment]
Is there no way to work around `os.PathLike` in #14515 to unofficially support Python 3.4 or 3.5 for some time?
But in general I don't think that `zipfile_deterministic.py` is often used by contributors. For me it would be no problem to use Python 3.6+ or Docker.

**Comment by Hainish** [comment]
@jeremyn

> Are you asking how many contributors would be seriously inconvenienced by requiring 3.6 for builds? In that case you might just ask that question, in here or #13815 ("Which developer operating systems do we support?"), and ping the maintainers and top 10 or so contributors for a thumbs-up/thumbs-down.

Yes, that's pretty much what I'm asking.  Pinging contribs @cschanaj @Bisaloo @Jochen-A-Fuerbacher @youdly @numismatika - apologies if you're not interested in this discussion or only use the in-extension debugging ux for ruleset creation, but this discussion may be relevant if you compile the extension.

> Is there no way to work around os.PathLike in #14515 to unofficially support Python 3.4 or 3.5 for some time?

I'm sure there is, but I'd like to avoid stepping from pyton 3.4 (or 3.5) to 3.6 for every migrated piece of infrastructure.

> But in general I don't think that zipfile_deterministic.py is often used by contributors. For me it would be no problem to use Python 3.6+ or Docker.

Thanks for your input!

**Comment by Bisaloo** [comment]
Python 3.6+ is fine for me.

(I am not sure I support the Docker suggestion though but that's another discussion. Just thought I would clarify this point since things can get complicated in such long threads.)

**Comment by jeremyn** [comment]
A Python 3.6 dependency is workable for me.

( @Bisaloo The "another discussion" for Docker is probably issue https://github.com/EFForg/https-everywhere/issues/13815 .)

**Comment by cschanaj** [comment]
I have been using the stable releases of Debian (currently Stretch) for development, its default Python 3 version is v3.5 as per https://packages.debian.org/stretch/python3. So I guess we should be using Python 3.4+ if we are planning to support the Debian platform. 

P.S. I am fine with Python 3.6 as a build dependency.

**Comment by Hainish** [comment]
Thank you @Bisaloo, @jeremyn, @cschanaj.  If there is no more feedback on platforms by the end of the day, I'll merge in #14515 and proceed on conversion of scripts to python 3.6.

**Comment by Hainish** [comment]
The main HTTPS Everywhere repository is now fully converted to python 3.6, unless I've missed something.  `install-dev-dependencies.sh` includes commands that still install Python 2.7 on non-debian based platforms, but Since our supported development OS will be standardized to Ubuntu 18.04 LTS in the future, we should be removing these instructions anyway.  This hasn't been discussed yet and is still up for debate, but I think it follows from #13815.

Still TODO:
1. Convert https://github.com/EFForg/https-everywhere-atlas to Python 3
2. Remove Python 2 dependencies from https://github.com/EFForg/https-everywhere-docker-base

**Comment by Hainish** [comment]
I've informed the Tor Browser team of changes in our build tooling, since they build HTTPS Everywhere from source for inclusion in TB.
